hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java

```diff
@@ -320,46 +330,58 @@ public void write(byte[] b, int off, int len) throws IOException {
     if (len == 0) {
       return;
     }
-
-    while (len > 0) {
-      allocateNewBufferIfNeeded();
-      final int writeLen = Math.min(currentBufferRemaining, len);
-      currentBuffer.put(b, off, writeLen);
-      currentBufferRemaining -= writeLen;
-      writeChunkIfNeeded();
-      off += writeLen;
-      len -= writeLen;
-      updateWrittenDataLength(writeLen);
-      doFlushOrWatchIfNeeded();
+    synchronized (this) {
+      while (len > 0) {
+        allocateNewBufferIfNeeded();
+        final int writeLen = Math.min(currentBufferRemaining, len);
+        currentBuffer.put(b, off, writeLen);
+        currentBufferRemaining -= writeLen;
+        updateWrittenDataLength(writeLen);
+        writeChunkIfNeeded();
+        off += writeLen;
+        len -= writeLen;
+        doFlushOrWatchIfNeeded();
+      }
     }
   }
 
-  public void updateWrittenDataLength(int writeLen) {
+  public synchronized void updateWrittenDataLength(int writeLen) {
```

Comment:
1. This can be `protected`
2. Looks like all 3 callers of this method are already calling this in `synchronized` block. But I guess it doesn't hurt.

```suggestion
  protected synchronized void updateWrittenDataLength(int writeLen) {
```

Reply:
Agree. I added synchronized because findbugs made me do that.